src/common/crc32c.cc

```diff
@@ -45,3 +45,54 @@ ceph_crc32c_func_t ceph_choose_crc32(void)
  */
 ceph_crc32c_func_t ceph_crc32c_func = ceph_choose_crc32();
 
+struct crc_turbo_struct {
+  uint32_t val[32][32];
+};
+
+/*
+ * Look: http://crcutil.googlecode.com/files/crc-doc.1.0.pdf
+ * Here is implementation that goes 1 logical step further,
+ * it splits calculating CRC into jumps of length 1, 2, 4, 8, ....
+ * Each jump is performed on single input bit separately, xor-ed after that.
+ */
+crc_turbo_struct create_turbo_table()
+{
+  crc_turbo_struct table;
+  for (int bit = 0 ; bit < 32 ; bit++) {
+    table.val[0][bit] = ceph_crc32c_sctp(1UL << bit, nullptr, 1);
+  }
+  for (int range = 1; range <32 ; range++) {
+    for (int bit = 0 ; bit < 32 ; bit++) {
+      uint32_t crc_x = table.val[range-1][bit];
+      uint32_t crc_y = 0;
+      for (int b = 0 ; b < 32 ; b++) {
+        if ( (crc_x & (1UL << b)) != 0 ) {
+          crc_y = crc_y ^ table.val[range-1][b];
+        }
+      }
+      table.val[range][bit] = crc_y;
+    }
+  }
+  return table;
+}
+
+static crc_turbo_struct crc_turbo_table = create_turbo_table();
+
+uint32_t ceph_crc32c_zeros(uint32_t crc, unsigned len)
+{
+  int range = 0;
+  while (len != 0) {
+    uint32_t crc1 = 0;
+    if ((len & 1) == 1) {
+      for (int b = 0; b < 32 ; b++) {
+        if ((crc & (1UL << b)) != 0 ) {
```

Comment:
Since you're not using value of "crc" after this loop, you could just do "if (crc % 1)" and "crc >> 1", then break out early when crc becames 0.